broadinstitute/single_cell_portal_core
# Adding history tracking to study and study file

This adds the[ `mongoid-history`](https://github.com/mongoid/mongoid-history) gem for tracking changes to Study and StudyFile objects.  Once this change is out and we're comfortable with it, I suspect we will want to add tracking to most of our models (e.g. User, BrandingGroup, but not DataArray).   The tricky bit is identifying which fields to track, so you don't end up with audit tables that grow too fast.

This also updates the nightly admin email with a listing of studies created, deleted, and updated.  Right now the display is rough, high level stuff.  We can see how useful it is and expand later.  Currently it's best interpreted as 'here are some things you may want to look at'.  The audit history itself is often complex and there are lots of idiosyncracies (e.g. a user update to a study description results in two update logs, since we have an after_save hook that strips tags, and reporting the real name of a deleted study qould have required more time than I had to go through the logs and get the name prior to it being overwritten with DELETE-XXX).

![image](https://user-images.githubusercontent.com/2800795/131567907-0cbd0e65-0118-42fe-8d5a-f86b31c3ae8a.png)

 I regret running out of time to implement a 'modification history' tab in the portal study view, but the modification history can be relatively easily explored in the rails console in the meantime to answer more detailed questions.  e.g. `HistoryTracker.where(:created_at.lt => 1.hour.ago)` will get all changes within the last hour, and `Study.find_by(accession: 'SCPXXX').history_tracks` will get the history for a given study

This PR also includes a change to the way we handle emails in development mode.  Before, development mode did not perform email delivery.  Now, development mode does send emails, but restricts them to only be sent to the developer's email.  This makes testing of email capabilities easier, and may help in making developers aware of when emails are triggered.

One other minor change -- the 'File integrity check' is moved to the top of the report if a discrepancy is detected

TO TEST:
 0. check out change, run bundle install
 1. Do some stuff to studies, update, delete, create, etc...
 2. In a rails console, run `SingleCellMailer.nightly_admin_report.deliver_now` 
 3. confirm email gets sent roughly reflecting what you did
